Let's talk about prealloc linter

### Your message to us

Original messages by @kevgo consolidated by @legeana for a discussion in the same thread.

https://github.com/git-town/git-town/pull/4511#issuecomment-2576150680

In my experience, the prealloc linter has been quite reliable—I’ve never encountered false positives with it. I recall reading that preallocating slices to their known size is considered a best practice, as it can reduce the number of allocations and minimize copying between backing arrays. Do you have any resources or links that explain why preallocating might be considered a premature optimization?

That said, for the Git Town codebase, this likely isn't a critical concern given the limited amount of data we handle. I'm fine with not using the prealloc linter if you prefer to leave it disabled.

https://github.com/git-town/git-town/pull/4403#issuecomment-2576253870

I’d like to request that we don't disable the prealloc linter across the entire repo. The end-to-end test suite currently calls the Git Town executable over 1600 times. At this scale, even small inefficiencies can compound, making it worthwhile to tackle some of the low-hanging fruit around performance.

For context, there was a time when running make cuke took 18 minutes. Through targeted, data-driven optimizations, we've managed to bring that down to just 22 seconds today (on Apple Silicon running Asahi Linux—macOS remains slower due to filesystem bottlenecks). This makes it possible to lean on the end-to-end tests as part of development.

Here are a couple of examples of optimizations that made a meaningful impact on the test suite’s performance:

- https://github.com/git-town/git-town/pull/1441 found that creating a Git repo from scratch for each test was a major bottleneck. The solution involved setting up a fully configured Git environment once at the start of the test suite and cloning it for each test.
- https://github.com/git-town/git-town/pull/1447 found that disabling the garbage collector significantly reduces test suite runtime on CI (which is a severely performance constrained environment).

I'm curious about how you see this, and if I'm missing something. Let me know if you’d like more background or additional details! 🙂

> Preallocating slices in Go is straightforward and non-intrusive, and we already have a linter to enforce it. While its immediate benefits might be small, it’s part of cultivating a disciplined approach to programming—like always closing files. It’s less about strict necessity and more about building a foundation of best practices.

Always closing files is not a fair analogy. Closing a file prevents a resource leak, or even affects correctness depending on a situation. Preallocation on the other hand is only a potential optimisation.

I would like to consider two situation.

## True Positive

The slice size is known in advance and prealloc's suggestion is trivially actionable. A change required is effectively:

```
- var x []T
+ x := make([]T, 0, N)
```

I think in this situation the preallocation very slightly increases the code complexity and very slightly reduces readability. The reader needs to know what `make` does and what its parameters are. For anyone experienced this should be negligible. I personally don't have strong feelings either way as this is basically a stylistic choice.

## False Positive

The slice size is not know in advance and prealloc's suggestion is not trivially actionable. This came up #4511 in the [tRunSubtestNames](https://github.com/git-town/git-town/blob/b2de6b1401471a990119a991a4d6d11c3a550d09/tools/tests_sorted/lint_func.go#L67-L92) function:

```
lint_func.go:68:2: Consider pre-allocating `subtests` (prealloc)
        var subtests []string
        ^
```

I found this unhelpful and proposed to disable the linter. Thinking more about it maybe another solution is to report the bug upstream as this change "magically" fixes the linter's concern:

```
diff --git a/tools/tests_sorted/lint_func.go b/tools/tests_sorted/lint_func.go
index f960e9549..cc612561b 100644
--- a/tools/tests_sorted/lint_func.go
+++ b/tools/tests_sorted/lint_func.go
@@ -83,10 +83,9 @@ func tRunSubtestNames(statements []ast.Stmt) ([]string, error) {
                        // This may indicate a linter bug.
                        return nil, err
                }
-               if !r.Success() {
-                       continue
+               if r.Success() {
+                       subtests = append(subtests, subtest)
                }
-               subtests = append(subtests, subtest)
        }
        return subtests, nil
 }
```

It is clear that the linter isn't correctly interpreting a conditional `continue`.

But this demonstrates my point nonetheless.
- The linter forces the developer to make a meaningless change to the code to satisfy a robot.
- Naive solution is to do a double scanning approach which greatly increases complexity.
- A more appropriate solution is to add `//nolint:prealloc // Unknown size.`, but that forces an unfamiliar developer to justify why they have the right to go against an established project style.

I worry that all the efficiency we gain in true positives we lose in a single false positive. Bad automation often feels worse than no automation. In my opinion such false positives fall under your "Harmful premature optimizations".

## Way forward

Fundamentally it is your decision on what approach git-town should take. I would consider the following:

- Document in [DEVELOPMENT](https://github.com/git-town/git-town/blob/main/docs/DEVELOPMENT.md#run-linters) how to disable incorrect linters. Possibly encourage a per-line or even per-file disablement of linters if they misbehave and deal with it in code review. This will mitigate most of the negative impact of false positives.
- Complain to prealloc owners about the incorrect behaviour.
- Disable `prealloc` in all `git-town`, or
- Enable `prealloc` in `tools/tests_sorted` to align it with the rest of the project and add `//nolint:prealloc` to `tRunSubtestNames`.

And just to make sure this came across right, it's not at all a big deal to me either way, this is incredibly minor in the grand scheme of things.

--- a/Makefile
+++ b/Makefile
@@ -74,7 +74,7 @@ lint-all: lint tools/rta@${RTA_VERSION}  # runs all linters
 	@echo lint tools/structs_sorted
 	@(cd tools/structs_sorted && ../rta golangci-lint run)
 	@echo lint tools/tests_sorted
-	@(cd tools/tests_sorted && ../rta golangci-lint run --disable prealloc)
+	@(cd tools/tests_sorted && ../rta golangci-lint run)
 	@echo lint tools/lint_steps
 	@(cd tools/lint_steps && ../rta golangci-lint run)
 

--- a/tools/tests_sorted/Makefile
+++ b/tools/tests_sorted/Makefile
@@ -3,6 +3,5 @@ help:  # prints all available targets
 
 test:  # runs all tests
 	go test ./...
-	# prealloc suggests a premature optimisation.
-	../rta golangci-lint run --disable prealloc
+	../rta golangci-lint run
 	../rta deadcode github.com/git-town/git-town/tools/tests_sorted

--- a/tools/tests_sorted/lint_func.go
+++ b/tools/tests_sorted/lint_func.go
@@ -65,7 +65,7 @@ func isTestFunc(funcType *ast.FuncType) bool {
 // tRunSubtestNames returns a list of subtest names from `t.Run(...)`
 // invocations.
 func tRunSubtestNames(statements []ast.Stmt) ([]string, error) {
-	var subtests []string
+	var subtests []string //nolint:prealloc // alexkohler/prealloc#16
 	testNameExtractor := &matcher.FirstStringArgFromFuncCallExtractor{
 		FuncMatcher: &matcher.IdentSelectorMatcher{
 			Namespace: "t",

